scoring/score.py

```diff
@@ -1,20 +1,64 @@
+# -*- coding: utf-8 -*-
+
+import itertools
 import collections
 
 
 class InvalidScoresheetException(Exception):
     pass
 
 
+POINTS_HELD = 1
+POINTS_GROUND = 2
+POINTS_PLATFORM = 5
+
+NUM_TOKENS = 16
+
+
 class Scorer(object):
     def __init__(self, teams_data, arena_data):
         self._teams_data = teams_data
         self._arena_data = arena_data
 
     def validate(self, extra):
-        pass
+        token_counts = tuple(itertools.chain(
+            [x.get('tokens_held', 0) for x in self._teams_data.values()],
+            [extra.get('tokens_ground', 0)],
+            [x.get('tokens_ground', 0) for x in self._arena_data.values()],
+            [x.get('tokens_platform', 0) for x in self._arena_data.values()],
+        ))
+
+        negative_counts = [x for x in token_counts if x < 0]
+        if negative_counts:
+            raise InvalidScoresheetException(
+                "Negative token counts are not valid (got {})".format(
+                    " and ".join(str(x) for x in negative_counts),
+                ),
+            )
+
+        num_tokens = sum(token_counts)
```

Comment:
Maybe we should check for fractional tokens?

Reply:
I'm not sure this is worth it. I'm reasonably sure that the scorer UI only allows integers.

Reply:
`assert num_tokens % 1 == 0`

Reply:
python3.7.3 if it makes a difference
```
>>> 3.0%1 == 0
True
```
if it is worth it, we could go even further and do `type(num_tokens) == type(1)`?

Reply:
c587c01